[epoll] Unix.select conversion: replace with stdext modules/calls or polly calls

edwintorok opened the pull request:
Builds on top of https://github.com/xapi-project/xen-api/pull/5704

edwintorok commented:
The previous PR for this was https://github.com/xapi-project/xen-api/pull/4877, we should check whether all the review comments from that PR got addressed.

lindig commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -547,7 +547,8 @@ let really_read_string fd length =
 
-let to_milliseconds ms = ms *. 1000. |> ceil |> int_of_float
+let to_milliseconds s =
+  Mtime.Span.to_float_ns s *. 1e-6 |> Float.ceil |> int_of_float
```
This is not the same functionality? The old code was confusing using `ms` as argument when it probably was seconds. What is the unit of the argument now? seconds, I assume. Why do we need `Mtime.Span.to_float_ns`?

edwintorok replied:
Unfortunately there is no way to directly convert an `Mtime.Span.t` to seconds. The only conversion we can do is to nanoseconds, which we then need to convert into the units that we desire (seconds, or milliseconds, in this case milliseconds, because that is what Polly expects).

psafont replied:
I don't think there's a need to use floats here (spans are always positive)
```suggestion
Int64.div (Mtime.Span.to_uint_ns s) 1_000_000L |> Int64.to_int
```

edwintorok replied:
Using int64 will be a little bit slower (it needs to allocate a boxed int64, both `int` and `float` are unboxed), but in this context it probably doesn't matter, and might be more accurate.

edwintorok replied:
Changed to use int64.

edwintorok replied:
Actually this doesn't work, because it doesn't round correctly, I'll fix it.

edwintorok replied:
This works, and I added a comment explaining why (I had the very same bug in the original float version, which is why I introduced `Float.ceil`):
```
+(* Need to round up when converting to milliseconds:
+   The span will have the amount of time already elapsed subtracted, so if we are asked to wait
+   for 1ms, the span will be slightly smaller than that.
+   If we truncate then all waits of 1ms will be 0ms instead, which would be wrong, as we wouldn't wait at all.
+   If we round to nearest rather than up, then waits of 1ms will be ~0.5ms instead, since values below 0.5ms will get
+   rounded to 0.
+   Rounding up works correctly, and makes the tests pass.
+ *)
+let to_milliseconds span =
+  Int64.div (Int64.add 999_999L @@ Mtime.Span.to_uint64_ns span) 1_000_000L |> Int64.to_int
```

lindig commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -586,6 +585,55 @@ let with_polly_wait kind fd f =
+        invalid_arg (Printf.sprintf "Span out of range: %gs" seconds)
+
+  (* nanosecond conversion doesn't roundtrip, so use only microseconds, which does *)
+  let to_string t = Mtime.Span.to_float_ns t *. 1e-9 |> Printf.sprintf "%.6f"
```
Is the comment above correct?

edwintorok replied:
Yes, I tried using nanoseconds, but multiplying and dividing by `1e-9` results in something different to the original. It could probably be fixed by careful use of rounding, although some of this happens inside Mtime.

lindig commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -586,6 +585,55 @@ let with_polly_wait kind fd f =
       f wait fd
 
+module Timeout = struct
+  type t = Mtime.Span.t
```
I think these modules need a comment what the main abstraction here is. A timeout is a time span?

psafont replied:
I don't understand why this module is needed at all. When using time, I've never had to convert from and to string, other than logging. If a conversion from a string is needed, first convert it to float, or from an uint64, and use a conversion that clearly states the units. Having conversion functions with implicit units opens possibility to misuse and should be avoided.

edwintorok replied:
The timeout is present in a configuration file, so we need to convert from a string, and due to how config file parsing is implemented in XAPI we must always provide a conversion to a string too.

edwintorok replied:
We could potentially move this somewhere else that deals with configuration files (although both xapi and xenopsd need it)

edwintorok replied:
> use a conversion that clearly states the units

I could require to always include `s` at the end, that'd probably be good, because we had timeout configurations in xapi that were sometimes in seconds, sometimes in milliseconds, so this should make it clearer.

psafont replied:
> The timeout is present in a configuration file, so we need to convert from a string,

But that is already taken care of by `Arg.Float` or `Arg.Int`

> and due to how config file parsing is implemented in XAPI we must always provide a conversion to a string too.

What is the to_string used for, exactly? Why is roundtripping being tested? What's stopping the printing from being defined as `Fmt.to_to_string Mtime.Span.pp`?

edwintorok replied:
Xcp_service.configure takes both a function that parses a string, and one that converts it back (e.g. to print in debug logs, or to print a configuration dump). That should be in a format that you can take and put into a config file, so it needs to be able to parse it back.
I could use Mtime.Span.pp, but there is no equivalent from_string function that can parse the pp output (the format of 'pp' is not defined on purpose, it is a pretty-printer, not a serializer).

*If* I can make all these timestamp be just simple floats expressed in seconds then none of this is needed (but I think there are some configuration values in XAPI where the values are not in seconds, but perhaps I can rename them and change them to be in seconds for consistency).

edwintorok replied:
Looking at the actual usage of Timer/Timeout so far it looks like what I was trying to protect against (accidentally using another value of type Mtime.Span.t instead of timeout) is a bit unlikely, so maybe I can drop this module (Timeout) and keep just Timer.

edwintorok replied:
After rebasing on master there is now only an Mtime.Span.t, and no Timeout anymore.

lindig commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -586,6 +585,55 @@ let with_polly_wait kind fd f =
+end
+
+module Timer = struct
+  let delay dt = Mtime.Span.to_float_ns dt *. 1e-9 |> Thread.delay
```
What is a timer? Some comment that introduces the abstraction.

edwintorok replied:
I'm adding some ocamldoc comments. Here is what I have so far:
```
+(** A "timeout" is a time span, relative to the start of an operation.
+  It is usually used to express a limit on the duration of an operation.
+
+  Not all time spans are timeouts, e.g. the difference between 2 timestamps (a duration) is not a timeout,
+  because it only expresses measured, elapsed time, not a limit.
+
+  All timeouts are time spans, e.g. a 30s timeout on a read operation means that we should raise an exception if the
+  read operation doesn't complete within 30s from the moment it started.
+ *)
 module Timeout : sig
   type t = private Mtime.Span.t
+  (** upper bound for the duration of an operation *)
```

```
+(** A timer is started at the same time as a particular operation.
+  It measures elapsed time from that moment, and has a desired upper bound, its timeout.
+
+  At any moment we can query the timer and get the remaining time, which is either
+   [Spare] if the timeout has not been reached yet,
+   or [Excess] if the timer has already expired, i.e. we have already exceeded the timeout.
+
+ *)
 module Timer : sig
   val delay : Mtime.Span.t -> unit
   (** [delay dt] delays the execution of the current thread for [dt]. *)

   type t
+  (** a timer that was started at a particular moment in time, and measures time relative to it *)

   type remaining = Spare of Mtime.Span.t | Excess of Mtime.Span.t

   val start : timeout:Timeout.t -> t
-  (** [start ~timeout] *)
+  (** [start ~timeout] starts a timer, with the relative deadline [timeout]. *)
```

Perhaps we should call 'Excess' as 'Expired' instead, it'd fit with the typical usage of a timer better.

edwintorok replied:
We are now using the Timer module from Clock which has equivalent documentation comments.

lindig commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.mli:
```diff
@@ -146,15 +146,54 @@ val try_read_string : ?limit:int -> Unix.file_descr -> string
 
-val time_limited_write : Unix.file_descr -> int -> bytes -> float -> unit
+module Timeout : sig
+  type t = private Mtime.Span.t
```
Again, introduce the abstraction in some words. It's a time span.

edwintorok replied:
module dropped.

lindig commented on ocaml/xapi-idl/lib_test/scheduler_test.ml:
```diff
@@ -37,7 +37,13 @@ let timed_wait_callback ~msg ?(time_min = 0.) ?(eps = 0.1) ?(time_max = 60.) f =
+      let finally () = Polly.close polly in
+      Fun.protect ~finally @@ fun () ->
+      let ready =
+        Polly.wait polly 1 (time_max *. 1000. |> int_of_float) (fun _ _ _ -> ())
```
We probably should give `Polly.wait` a named argument that includes the unit for the delay.

edwintorok replied:
This seems like a very complex way of doing Unix.sleepf, or I've accidentally dropped the file descriptor (again!).
And indeed I have, there should've been an `rd` here.

I'm defining a `wait_timed_read` in unixext to avoid this (and the other ) similar ubg.

lindig commented on ocaml/xapi/xapi_globs.ml:
```diff
@@ -951,7 +951,11 @@ let ignore_vtpm_unimplemented = ref false
+type xapi_globs_spec_ty =
+  | Float of float ref
+  | Int of int ref
+  | Timeout of Xapi_stdext_unix.Unixext.Timeout.t ref
```
What is the difference between a span and a timeout? What are Float and Int denoting?

edwintorok replied:
Float and Int are the existing types for xapi globs entries. Timeout is the new type. Some ocamldoc comments would probably be useful, I'll add some.

psafont replied:
Not all the time spans in xapi globs are for timeouts, I believe TimeSpan would encompass periods as well.  
```
| TimeSpan of Mtime.Span.t ref
```

psafont replied:
I don't understand why Span and Timeout need different variants here either

edwintorok replied:
I should probably change `remaining ?attempt_delay` to take a `Timeout.t` argument, then I can drop `Span` from here.

edwintorok replied:
Rebased on master, there is now a short and long duration instead (and we only use the short duration).

edwintorok commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml:
```diff
@@ -64,22 +68,26 @@ let test_time_limited_write =
       let open Observations in
-      let elapsed_s = Mtime.Span.to_float_ns !test_elapsed *. 1e-9 in
-      if elapsed_s > timeout +. 0.5 then
+      let elapsed = !test_elapsed in
```
`elapsed` is an `Mtime.Span.t`, which I think internally is an integer.

psafont replied:
Internally, they are int64. In any case `Mtime.Span.equal` should do the right thing.

edwintorok replied:
THis will use Mtime.Span.compare  in the new branch.

psafont replied:
Clock.Timer provides span_is_shorter and span_is_longer, which are easier to understand than compare

edwintorok replied:
It is easier to check that I haven't introduced a bug when changing from `>` to Mtime.Span.compare than by using one of these other names.
E.g.:
```
-      if elapsed_s > timeout +. 0.5 then
+      let elapsed = !test_elapsed in
+      let timeout_extra =
+        Mtime.Span.(add (timeout :> Mtime.Span.t) @@ (500 * ms))
+      in
+      if Mtime.Span.compare elapsed timeout_extra > 0 then
```

Here I need to check that I have the arguments in the same order, and the same kind of comparison operator, so 
`elapsed_s > timeout..` becomes `Mtime.Span.compare elapsed timeout.. > 0`.

With is_shorter or is_longer it is very easy to use the wrong one and it won't be that obvious from `git diff` that I got it wrong.

I agree that if I was writing brand new code then `is_shorter` or `is_longer` would've been easier to use.

edwintorok commented on ocaml/database/master_connection.ml:
```diff
@@ -171,7 +171,10 @@ let open_secure_connection () =
+  let polly = Polly.create () in
+  let finally () = Polly.close polly in
+  Fun.protect ~finally @@ fun () ->
+  let fd_closed = Polly.wait polly 1 5000 (fun _ _ _ -> ()) > 0 in
```
Good spot, the fd was completely lost here, so Polly was always waiting until a timeout here. It means that it is now no longer able to detect an stunnel connection that dies too quickly.

I should probably add a unit test to detect this in addition to fixing it.

edwintorok replied:
Maybe a less error-prone way would be to write an equivalent of `Thread.wait_timed_read` implemented using Polly, instead of select. Then we can unit-test that, and use if everywhere, instead of making subtle mistakes that might go undetected.

edwintorok replied:
I've introduced a `wait_timed_read` in Unixext, and replaced all these direct Polly calls with that one.

psafont commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -586,6 +586,55 @@ let with_polly_wait kind fd f =
+
+  let start ~timeout = {elapsed= Mtime_clock.counter (); timeout}
+
+  let elapsed {elapsed; _} = elapsed
```
Users shouldn't be given access to the counter, IMO:
```suggestion
  let elapsed t = Mtime_clock.count t.elapsed
```

edwintorok replied:
The function is not exported, its only user is `pp`. I can move it inside `pp` to make that clearer.

psafont commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.mli:
```diff
@@ -146,15 +146,54 @@ val try_read_string : ?limit:int -> Unix.file_descr -> string
+  val start : timeout:Timeout.t -> t
+  (** [start ~timeout] *)
+
+  val remaining : ?attempt_delay:Mtime.Span.t -> t -> remaining
```
I think the optional parameter should be removed altogether. Instead, a function to transform the timer should be added that can be used to the same effect:

```
val shorten_by : Mtime.Span.t -> t -> t
```
At that point we can also add the analogous function to extend it:
```
val extend_by : Mtime.Span.t -> t -> t
```

psafont commented on ocaml/xapi/xapi_globs.ml:
```diff
@@ -1114,19 +1122,38 @@ let xapi_globs_spec =
+let timeout ref str = ref := Xapi_stdext_unix.Unixext.Timeout.of_string str
+
+let span ref str =
+  ref := (Xapi_stdext_unix.Unixext.Timeout.(of_string str) :> Mtime.Span.t)
+
 let options_of_xapi_globs_spec =
   List.map
     (fun (name, ty) ->
       ( name
-      , (match ty with Float x -> Arg.Set_float x | Int x -> Arg.Set_int x)
+      , ( match ty with
+        | Float x ->
+            Arg.Set_float x
+        | Int x ->
+            Arg.Set_int x
+        | Timeout x ->
+            Arg.String (timeout x)
+        | Span x ->
+            Arg.String (span x)
```
I'd prefer to keep using Arg.Float here, I don't like having the conversion being made deep in the stack call.

```suggestion
let span ref x =
  let t = Float.to_int (1000. *. x) in
  ref := Mtime.Span.(t * ms)

let options_of_xapi_globs_spec =
  List.map
    (fun (name, ty) ->
      ( name
      , (match ty with Float x -> Arg.Set_float x | Int x -> Arg.Set_int x)
      , ( match ty with
        | Float x ->
            Arg.Set_float x
        | Int x ->
            Arg.Set_int x
        | Timeout x ->
            Arg.String (timeout x)
        | Span x ->
            Arg.Float (span x)
```

edwintorok commented:
I've rebased this on top of latest feature perf + master that includes the new clock module: https://github.com/edwintorok/xen-api/tree/private/edvint/epoll3-pr-updated.
I'll look at addressing the other review comments next.

edwintorok commented on ocaml/database/block_device_io.ml:
```diff
@@ -731,6 +735,8 @@ let dump = ref false
 
 let empty = ref false
 
+let target_response_time delta = Unixext.Timer.start ~timeout:delta
```
In the updated branch this is using `Clock.Timer.start ~duration:delta` now.

edwintorok commented on ocaml/libs/ezxenstore/core/watch.ml:
```diff
@@ -31,41 +31,25 @@ let has_fired ~xs x =
     true
   with Xs_protocol.Eagain -> false
 
+module Delay = Xapi_stdext_threads.Threadext.Delay
```
I've dropped the Timeout module, so now we only have Delay.

edwintorok commented on ocaml/database/db_globs.ml:
```diff
@@ -1,21 +1,31 @@
 (* Database globals *)
 
+open Xapi_stdext_unix
```
dropped the extra Timeout conversions.

edwintorok commented:
This PR depends on https://github.com/xapi-project/xen-api/pull/5766, which in turn depends on the quicktest bugfix on master (I've reverted the commit in this branch, to be able to test it, but I should eventually drop that revert once we got the proper fix).

edwintorok commented:
There were a lot of changes, rebases here to keep up with the changes on master and I found one rebase error so far: 
```
wait remaining_time;
wait remaining_time
```

I tried modifying the tests to detect this doubling, but it doesn't yet.

Putting this back to draft, also we'll need to carefully review the code, it should be simpler, but there were a lot of rebases/fixups and might've completely broken it (I haven't looked at the diffs yet, just at unit test results, and running a BVT now).

edwintorok commented:
Found the bug with HA: when removing the `Timeout` module and replacing it with the `Clock.Timer` module and spans there was no `delay` function, so I implemented one, but got the conversion wrong, and gave `Unix.sleepf` and argument that was in nanoseconds instead of seconds, so XAPI ended up waiting for a very long time before connecting to `block_device_io`, which gave up, raised a timeout error and exited with code 2.

I've now added a test for the `delay` function, and used the `Clock.Timer.span_to_s` to perform the conversion.

psafont commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml:
```diff
@@ -187,7 +187,97 @@ let test_proxy =
+let spans =
+  Gen.oneofa ([|1; 100; 300|] |> Array.map (fun v -> Mtime.Span.(v * ms)))
+
+let test_timer_remaining =
```
This test is now part of Clock.Timer: https://github.com/xapi-project/xen-api/blob/master/ocaml/libs/clock/test_timer.ml#L84

edwintorok replied:
Thanks, I dropped the code from here in a fixup, we can squash that once I rebase on latest master.

psafont commented on ocaml/database/redo_log.ml:
```diff
@@ -265,26 +263,30 @@ let connect sockpath latest_response_time =
+            "Waiting to connect to I/O process via socket %s (error was %s: \
+             %s)..."
+            sockpath b (Unix.error_message a) ;
+          D.debug "Remaining time for retries: %s"
+            (Fmt.to_to_string Mtime.Span.pp remaining) ;
```
```suggestion
          D.debug "Remaining time for retries: %a"
            Debug.Pp.mtime_span remaining) ;
```

edwintorok replied:
Applied (although manually, I think if I use the button here then DCO will complain)

psafont commented on ocaml/database/redo_log.ml:
```diff
@@ -265,26 +263,30 @@ let connect sockpath latest_response_time =
+          Unixext.delay_span attempt_delay ;
+          attempt ()
+      | Expired overrun ->
+          D.debug "%s timeout reached (%ss overrun)" __FUNCTION__
+            (Fmt.to_to_string Mtime.Span.pp overrun) ;
```
```suggestion
          D.debug "%s timeout reached (%a overrun)" __FUNCTION__
            Debug.Pp.mtime_span overrun ;
```

edwintorok replied:
Applied, this also fixes the extra `s` at the end which Mtime's printer might've already added.

lindig commented:
> Found the bug with HA: when removing the `Timeout` module and replacing it with the `Clock.Timer` module and spans there was no `delay` function, so I implemented one, but got the conversion wrong, and gave `Unix.sleepf` and argument that was in nanoseconds instead of seconds, so XAPI ended up waiting for a very long time before connecting to `block_device_io`, which gave up, raised a timeout error and exited with code 2.
> 
> I've now added a test for the `delay` function, and used the `Clock.Timer.span_to_s` to perform the conversion.

This perfectly demonstrates the problem of having the various units for time spans and why we have to minimise this and probably need to encode the expected unit in a label for arguments.

edwintorok commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml:
```diff
@@ -187,7 +199,103 @@ let test_proxy =
+  in
+
+  (* 3*half > timeout, so we expect Excess to be reported now *)
+  Unix.sleepf (2. *. half) ;
```
There are 2 sleeps: `Unix.sleep half` followed by `Unix.sleep (2. *. half)`.
Although this code has now moved to the `test_timer.ml` file, and I'm dropping it from this PR.

edwintorok commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml:
```diff
@@ -187,7 +199,103 @@ let test_proxy =
+          Mtime.Span.pp t Unixext.Timeout.pp timeout Mtime.Span.pp actual
+          Unixext.Timer.pp timer
+    | Spare t ->
+        if Mtime.Span.compare Mtime.Span.(2 * t) (timeout :> Mtime.Span.t) > 0
```
Changed, although it requires adding a 3rd branch with `()` to handle the `else` case.

edwintorok commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml:
```diff
@@ -64,22 +68,26 @@ let test_time_limited_write =
-          if elapsed_s < timeout then
-            Test.fail_reportf "Timed out earlier than requested: %f < %f"
-              elapsed_s timeout ;
+          let elapsed = !test_elapsed in
```
See https://github.com/xapi-project/xen-api/pull/5705#discussion_r1646251157

edwintorok commented on ocaml/libs/http-lib/buf_io.ml:
```diff
@@ -79,14 +79,17 @@ let is_full ic = ic.cur = 0 && ic.max = Bytes.length ic.buf
+        n
+      with Unix.Unix_error (Unix.(EAGAIN | EWOULDBLOCK), _, _) -> -1
+    in
+    Unix.setsockopt_float ic.fd Unix.SO_RCVTIMEO 0. ;
```
I'm a bit undecided on this one, there is an exception handler right above it,  and using `Fun.protect` does have an overhead (it needs to allocate the closure for `finally`).
OTOH it doesn't handle *all* errors, just one particular one, so other exceptions may still escape.

I changed it to `Fun.protect` to be safer.

edwintorok commented:
> This perfectly demonstrates the problem of having the various units for time spans and why we have to minimise this and probably need to encode the expected unit in a label for arguments.

I've attempted to hide these conversions inside the `Unixext` module, so code in XAPI shouldn't need to interact with this directly anymore:
* for epoll the conversion from Mtime.Span.t to milliseconds is handled internally in Unixext, XAPI passes a type safe Mtime.Span.t
* for Unix.sleepf again the call is hidden in Unixext, exposing only a type-safe Mtime.Span.t

`float`s can be ambigous. The usual convention in the Unix module is that floats represent seconds, and that is why I keep making this mistake.
But Mtime doesn't follow that convention and returns a float the represents nanoseconds. To be fair the name of the function is very explicit that it returns nanoseconds.

In general Mtime.Span.t is what we should use, and we should hide the conversions to other units needed by syscalls/library calls in small functions/modules that are *tested*. 
Testing  is the part that I was missing (I think it was tested before, but I dropped those tests when removing the `Timeout` module), but I've added it back now.

edwintorok commented:
I think this PR is "ready", but I'm waiting for `feature/perf` to be merged to master before undrafting it as that will require some updates to the target branch of the PR.

edwintorok commented:
Apparently there is a problem with the timeout tests now, will investigate tomorrow:
```
Timed out earlier than requested: 187μs < 1ms
       Raised at QCheck2.Test.fail_report in file "src/core/QCheck2.ml", line 1611, characters 22-41
       Called from Dune__exe__Unixext_test.test_time_limited_write.(fun) in file "ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml", line 87, characters 12-131
Error: Exception: qcheck: user fail:
Timed out earlier than requested: 98.5μs < 1ms
       Raised at QCheck2.Test.fail_report in file "src/core/QCheck2.ml", line 1611, characters 22-41
       Called from Dune__exe__Unixext_test.test_time_limited_write.(fun) in file "ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml", line 87, characters 12-131
Error: Exception: qcheck: user fail:
Timed out earlier than requested: 109μs < 1ms
       Raised at QCheck2.Test.fail_report in file "src/core/QCheck2.ml", line 1611, characters 22-41
       Called from Dune__exe__Unixext_test.test_time_limited_write.(fun) in file "ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml", line 87, characters 12-131
[✗]   44    0    1   22 /  100     0.4s Dune__exe__Unixext_test.test_time_limited_write
```

edwintorok commented:
Fixed a few more bugs due to differences between select/epoll with a 0 timeout, and SO_RCVTIMEO with a 0 timeout.
The former times out immediately, the latter never times out (0 = no timeout).

I've added some code to detect calls to setsockopt with a timeout <1e-6 and made it raise invalid_arg to catch these (similar to what I've done with the tests on master). There is one exception: when we finish a function, we "restore" the timeout to 0, so in that case the call is allowed (I inserted the check into `with_socket_timeout` rather than `set_socket_timeout`).

There was in fact already a wrapper in the Http module for this, I moved it to unixext so we can reuse it.

psafont commented:
The PR includes commits from master, only the latest commits should be reviewed (all authored by Edwin)

psafont commented on ocaml/libs/ezxenstore/core/dune:
```diff
@@ -9,5 +9,6 @@
     (re_export xenstore)
     (re_export xenstore_transport)
     threads.posix
+    xapi-stdext-threads
```
There's a TODO in this commit, remove it, explaining the fix. I don't it should have a todo regarding upstreaming

edwintorok replied:
https://github.com/xapi-project/ezxenstore/pull/19

psafont commented on ocaml/database/block_device_io.ml:
```diff
@@ -325,16 +325,20 @@ let listen_on sock =
   s
 
 let accept_conn s latest_response_time =
-  let now = Unix.gettimeofday () in
```
Commit contains:
> TODO: needs some extra testing of block_device_io, which doesn't many suitable tests.

Has this been done in some way?

edwintorok replied:
There is 1 XenRT test capable of detecting breakages in the redolog, the one that got added for HA vTPM testing.
(all other HA XenRT tests used to pass with a completely broken redolog in the past).

There is a test for epoll on block devices part of the stdext quickcheck tests, although it requires root so it is not run in the CI.
I'll add some code to the XAPI quicktests so we can run some of these fuzzers in Dom0.

I think it is still worthwhile trying to add a separate test for the redolog itself that runs as part of the quicktests (it needs a blockdevice, and setting one up even in loopback mode requires root, so can't be done in the CI).

psafont commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/test/unixext_test.ml:
```diff
@@ -64,22 +67,25 @@ let test_time_limited_write =
-            Test.fail_reportf "Timed out earlier than requested: %f < %f"
-              elapsed_s timeout ;
+          let elapsed = !test_elapsed in
+          if Mtime.Span.compare elapsed timeout < 0 then
```
```suggestion
          if Clock.Timer.span_is_shorter elapsed ~than:timeout then
```

psafont commented on ocaml/libs/xapi-stdext/lib/xapi-stdext-unix/unixext.ml:
```diff
@@ -633,44 +642,57 @@ let time_limited_read filedesc length target_response_time =
+        if not (wait remaining_time) then raise Timeout ;
+        let bytes_to_read = total_bytes_to_read - !bytes_read in
+        let bytes =
+          try Unix.read filedesc buf !bytes_read bytes_to_read
```
Code seems almost all duplicated, only the function called changes (write / read), can this be factored out?

edwintorok replied:
Xapi-fdcaps has a better factored out implementation (there is a `repeat_read` and `repeat_write`, although much of the code of those 2 functions could be shared too, and there is no Timer version, but one could be easily introduced).

Xapi-fdcap is what I use to test unixext, so if we switch unixext over to xapi-fdcaps directly, then it would be testing itself with itself (although the way timeout testing is done is independent of how it is implemented, so the test would still be a useful one).

Unixext could be changed to use xapi-fdcaps, although the type changes will probably ripple through the codebase. It might result in an overall safer code, because we'd know from its type whether a file descriptor is a socket or not for example (and then whether calling setsockopt is a valid call or not).

I'll check whether switching over directly to fdcaps would result in simpler code, if not I'll factor out the code here in stdext.

edwintorok commented:
I should do a separate update from master PR again to make this easier to review.

edwintorok commented:
I'm going to take a different approach with epoll, see https://github.com/xapi-project/xen-api/compare/master...edwintorok:xen-api:private/edvint/epoll4.
That should be a much smaller change (still needs some tests) + a mechanical search/replace on the codebase, and then we can do the other improvements from this branch on top of it (I didn't want to do it that way originally because introduces a performance problem, but we've got the rest of the commits now that avoid it, it just taking too long review them).

edwintorok commented:
I've split this PR:

PRs that should go to master (to avoid further conflicts on the epoll branch, and don't yet convert to epoll):
https://github.com/xapi-project/xen-api/pull/5861
https://github.com/xapi-project/xen-api/pull/5862
https://github.com/xapi-project/xen-api/pull/5863

Alternative approach started here (untested so far):
https://github.com/xapi-project/xen-api/pull/5864

We can then do the Timer and xapi-fdcaps conversion on top.
